include/swift/AST/DiagnosticEngine.h

```diff
@@ -564,6 +568,12 @@ namespace swift {
     /// results that we can point to on the command line.
     llvm::DenseMap<const Decl *, SourceLoc> PrettyPrintedDeclarations;
 
+    llvm::BumpPtrAllocator TransactionAllocator;
+    /// A set of all strings involved in current transactional chain.
+    /// This is required because diagnostics are not directly emitted
+    /// but rather stored until all transactions complete.
+    llvm::StringMap<char, llvm::BumpPtrAllocator &> TransactionStrings;
```

Comment:
Clever way to avoid duplicates and to not have to directly talk to BumpPtrAllocator. You can take it a step further by using StringSet, and then you don't need the ignored `char`.

Reply:
This is not my idea, I took it from `ASTContext`, but thanks! I didn't use `StringSet` because comment in `ASTContext` mentioned that it doesn't support passing allocators, but now that I looked at it, apparently it does, so I'm going to switch to that!

Reply:
So yeah, unfortunately there is no constructor in `StringSet` which takes allocator, so I'd have to go with `StringMap` here still.